### PR TITLE
64 bit interrupts

### PR DESCRIPTION
**Pull request by DrDeano**
Add 64 bit interrupts to handle IRQs and ISR exceptions.
Refactored the x86 architecture code for common code.
Part of #169

### CODE REVIEW

File: src/kernel/arch/x86/32bit/arch.zig [hunk 1]
```diff
@@ -419,12 +391,14 @@ pub fn getDateTime() DateTime {
 ///                                         paging.
 ///
 pub fn init(mem_profile: *const MemProfile) void {
-    gdt.init();
-    idt.init();
+    gdt_common.init(gdt_32.Tss, &gdt_32.tss_entry);
+    idt_common.init(idt_32.IdtEntry, &idt_32.table);
+
+    _ = interrupts;
```
Comment (DrDeano) [inline src/kernel/arch/x86/32bit/arch.zig:399]: This line is needed to reference the `getInterruptStub()` function, unless there is a better way of doing it.

File: src/kernel/arch/x86/64bit/arch.zig [hunk 2]
```diff
@@ -332,15 +338,23 @@ pub fn initMem(boot_payload: BootPayload) Allocator.Error!MemProfile {
...
 ///     IN mem_profile: *const MemProfile - The memory profile of the computer. Used to set up
 ///                                         paging.
 ///
 pub fn init(mem_profile: *const MemProfile) void {
-    gdt.init();
-    idt.init();
+    gdt_common.init(gdt_64.Tss, &gdt_64.tss_entry);
+    idt_common.init(idt_64.IdtEntry, &idt_64.table);
+
+    _ = interrupts;
```
Comment (DrDeano) [inline src/kernel/arch/x86/64bit/arch.zig:353]: This line is needed to reference the `getInterruptStub()` function, unless there is a better way of doing it.

File: src/kernel/arch/x86/32bit/interrupts.zig [hunk 3]
```diff
@@ -1,48 +1,28 @@
...
+        \\mov   %[offset], %%ds
+        \\mov   %[offset], %%es
+        \\mov   %[offset], %%fs
+        \\mov   %[offset], %%gs
         \\mov   %%esp, %%eax
         \\push  %%eax
         \\call  handler
         \\mov   %%eax, %%esp
+        :
+        : [offset] "rm" (gdt_common.KERNEL_DATA_OFFSET)
```
Comment (SamTebbs33) [inline src/kernel/arch/x86/32bit/interrupts.zig:25]: Much cleaner :)

File: src/kernel/arch/x86/64bit/arch.zig (hunk 2 above)
Comment (SamTebbs33) [reply src/kernel/arch/x86/64bit/arch.zig:353]: I suggest adding a comment explaining that so we don't forget why it's there

File: src/kernel/arch/x86/64bit/interrupts.zig [hunk 4]
```diff
@@ -0,0 +1,59 @@
+const gdt_common = @import("../common/gdt.zig");
+
+///
+/// The common assembly that all exceptions and interrupts will call.
+///
+export fn commonStub() callconv(.Naked) void {
+    // Calling convention for x86_64 is for first parameter is in RDI (not the stack as for x86 (32 bit))
+    asm volatile (
+        \\push %%rdx
```
Comment (SamTebbs33) [inline src/kernel/arch/x86/64bit/interrupts.zig:9]: That's a lot of registers

File: src/kernel/arch/x86/32bit/arch.zig (hunk 1 above)
Comment (SamTebbs33) [reply src/kernel/arch/x86/32bit/arch.zig:399]: Comment here would be good too.

**Review by SamTebbs33 — APPROVED** [review]
Noice

File: src/kernel/arch/x86/64bit/interrupts.zig (hunk 4 above)
Comment (DrDeano) [reply src/kernel/arch/x86/64bit/interrupts.zig:9]: ik right, welcome to 64 bit land

**Pull request merged**